CenterForOpenScience/waterbutler
# [ENG-302] Handle GitHub Rate Limiting

## Ticket

https://openscience.atlassian.net/browse/ENG-302

## Purpose

This PR replaces https://github.com/CenterForOpenScience/waterbutler/pull/292

h/t to @TomBaxter for the core part implement which is fully functional, I just polished the code

GitHub API has a [rate-limiting](https://developer.github.com/v3/#rate-limiting) feature that only allow 5000 requests per hour. The limit is not a **rate** but a **count**.  To handle this limitation, this PR enables WB to rate-limiting GitHub move/copy requests by allowing 1-2 requests per second.

### GitHub VS WB

* GitHub rate-limiting is enforced per user. For example, API requests made to GitHub by non-OSF products users have authorized **count** against the quota.

```
For API requests using Basic Authentication or OAuth, you can make up to 5000 requests
per hour. Authenticated requests are associated with the authenticated user, regardless
of whether Basic Authentication or an OAuth token was used. This means that all OAuth
applications authorized by a user share the same quota of 5000 requests per hour when
they authenticate with different tokens owned by the same user.
```

* WB rate-limiting is enforced per provider instance. However, there are a few features that actively preventing it from hitting the rate-limit cap when multiple instances are present and when users are making a reasonable amount of requests using GitHub API outside OSF and WB.
  * Reserve a portion (default is 10%) of the pool for requests that does not go through WB's rate-limiting and for multiple instances scenario mentioned below.
  * Actively (default is 1 minute) updates the request rate based the response headers GitHub returns.
  * Slow down the request rate to 1 per 100 seconds if reaching the limit.

### What Will Work

If user tries to copy two or more repos of more than 2000 files at the same time, WB rate-limit can still work because none of them can make all the requests at the same time. The request rate limit gets refreshed every minute to reflect the most updated info from GitHub.

### What Will Not Work

As mentioned the request rate limit gets updated every minute. Within in that limit each instance makes request on their own pace assuming they are the only one using the quota. However, if users have other non-OSF apps that suddenly used up the quota within that minute, WB will fail.

## Changes

This is the original commit message (slightly modified) by Tom.

This PR only rate limits the GitHub provider's `.make_requests()` that run in celery tasks. Currently, the only tasks are "move" and "copy". The distinction is made by adding a `task_id` to the providers after they are instantiated in the task definition.

This PR uses the rate limit information returned in **all** GitHub response headers to limit the number of `.make_requests()` calls in such a way that the GitHub rate limit is never reached. In practice this usually results in a rate limit between 1-2 calls a second.

As a bonus, this PR adds a `GitHubRateLimitExceeded` exception, that will be checked for with each `.make_request()`. This exception ~~should never~~ is not expected to be raised during a move/copy celery task unless users suddenly use up the quota outside OSF or WB. This also helps non-celery requests as well.

## Side effects

GitHub move/copy request are rate-limited to 1-2 per second, no matter how many/few files are there. 

## QA Notes

Here is a dedicated [repo](https://github.com/cslzchen/OSF-Addon-Folder/tree/develop/feature-tasks/rate-limit) I created for test. Pease fork it for testing purpose. It has folders containing a variety number of files.

* Stage one, use the first 4 folders (files-[00128|00256|00512|01024]) to test and feel the side effect of speed decrease because of the rate limit.
* Stage two, [files-16384](https://github.com/cslzchen/OSF-Addon-Folder/tree/develop/feature-tasks/rate-limit/files-16384) contains 4 folders, each with 4096 files. Any of them should fail on `Prod` but  should succeed on `Staging`. Locally, it takes 2.5 hours to finish each one.
* Final stage, try copy 2 or more folders at the same time, each with a `different` OSF user. Please expect around 6 - 12 hours and it needs a stable `Staging`. You may want to skip this one.

## Deployment Notes

TBD
